CONTRIBUTING.md

```diff
@@ -0,0 +1,513 @@
+
+In addition, the PR submission should **answer the prompts of the Github template** we use for PRs.
+
+The code itself should adhere to our coding guidelines and conventions, which both submitters and reviewers should check.
+
+# Coding Conventions
+
+### Simplicity of implementation
+
+The most important consideration when accepting or rejecting a contribution is
+the simplicity (i.e. ease of understanding) of its implementation.
+Contributions that are "clever" or introduce functionality beyond the scope of
+the immediate problem they are meant to solve will be rejected.
+
+#### Type simplicity
+
+Simplicity of implementation includes simplicity of types.  Type parameters
+and associated types should only be used if there are at
+least two possible implementations of those types.
+
+Lifetime parameters should only be introduced if the compiler cannot deduce them
+on its own.
+
+### Builds with a stable Rust compiler
+
+We use a recent, stable Rust compiler.  Contributions should _not_
+require nightly Rust features to build and run.
+
+### Minimal dependencies
+
+Adding new package dependencies is very much discouraged.  Exceptions will be
+granted on a case-by-case basis, and only if deemed absolutely necessary.
+
+### Minimal global macros
+
+Adding new global macros is discouraged.  Exceptions will only be given if
+absolutely necessary.
+
+### No compiler warnings
+
+Contributions should not trigger compiler warnings if possible, and should not
+mask compiler warnings with macros.
+
+### Minimal `unsafe` code
+
+Contributions should not contain `unsafe` blocks if at all possible.
+
+# Coding Guidelines
+
+## Documentation
+
+* Each file must have a **copyright statement**.
+* Any new non-test modules should have **module-level documentation** explaining what the module does, and how it fits into the blockchain as a whole.
+* Any new files must have some **top-of-file documentation** that describes what the contained code does, and how it fits into the overall module.
+
+Within the source files, the following **code documentation** standards are expected:
+
+* Each public function, struct, enum, and trait should have a Rustdoc comment block describing the API contract it offers.  This goes for private structs and traits as well.
+* Each _non-trivial_ private function should likewise have a Rustdoc comment block.  Trivial ones that are self-explanatory, like getters and setters, do not need documentation.  If you are unsure if your function needs a docstring, err on the side of documenting it.
+* Each struct and enum member must have a Rustdoc comment string indicating what it does, and how it is used.  This can be as little as a one-liner, as long as the relevant information is communicated.
+
+## Factoring
+
+* **Public or exported struct, enum, and trait definitions go into the `mod.rs` file**.  Private structs, enums, and traits can go anywhere.
```

Comment:
Should a directory containing only a mod.rs be refactored into a single file?

```
foo
├── bar
│   └── mod.rs
└── baz.rs
```
should become:
```
foo
├── bar.rs
└── baz.rs
```

Reply:
I'm leaving discussion of `mod.rs` out of this file for now. But yes, generally single `mod.rs` is bad, and I'd discourage it in a PR.